pkg/ctxlock/watcher.go

```diff
@@ -0,0 +1,35 @@
+package ctxlock
+
+// A Watcher waits on two cancellation sources and makes sure to call the
+// wrapped function as soon as possible.
+type watcher struct {
+	onCancel func()
+	done     chan struct{}
+}
+
+func newWatcher(onCancel func()) *watcher {
+	return &watcher{
+		onCancel: onCancel,
+		done:     make(chan struct{}),
+	}
+}
+
+// Watch on the provided channel.
+func (w *watcher) Watch(ch <-chan struct{}) {
+	if ch == nil {
+		panic("nil channel")
+	}
+
+	select {
+	case <-ch:
+		w.onCancel()
+		<-w.done
+	case <-w.done:
+	}
+}
+
+// Unwatch tears down the watch. It should be called unconditionally.
+func (w *watcher) Unwatch() {
+	w.onCancel()
```

Comment:
Seems possible (likely) that `onCancel` will be called twice if I'm reading correctly, this should be documented if so

Reply:
Yeah, this is implicitly depending on the behavior specified in the `context` docs:

> A CancelFunc tells an operation to abandon its work. A CancelFunc does not wait for the work to stop. A CancelFunc may be called by multiple goroutines simultaneously. After the first call, subsequent calls to a CancelFunc do nothing.

I'll add a note and double-check that the `unlock` call does the right thing.